### ISSUE TITLE
Proposal: Nagios-compatible monitoring Semantic Conventions

### ISSUE THREAD

**Issue by martialblog** [open-telemetry/semantic-conventions#3148]
### Area(s)

area:new

### Propose new conventions

Hello,

I'd like to propose the inclusion of Semantic Conventions for Nagios-compatible monitoring data.

Nagios-compatible monitoring applications are tools that use "monitoring plugin" that conform to the 
"Plugin Output for Nagios" [1]. Example applications are Nagios, CheckMK, Icinga, Naemon, Shinken, Sensu.

These tools use "plugins" that are executed in intervals to determine the "state" of a host/service/component. There is vast ecosystem of these plugins [2] [3].

With this "state" - meaning the POSIX return code -the plugins can also return optional metrics (so called "performance data"). For example:

```
PING ok - Packet loss = 0%, RTA = 0.80 ms | percent_packet_loss=0, rta=0.80

DISK CRITICAL - free space: / 13969 MB (77% inode=99%); /boot 2 MB (0% inode=96%); 
| /=4151MB;14496;16308;0;18121 /boot=294MB;236;266;0;296
```

The performance data is provided in the following format [1] [4] [5]:

'label'=value[UOM];[warn];[crit];[min];[max]

The optional “warn” and “crit” denote thresholds for the value (meaning the threshold when they change from the OK return code to the Warning or Critical return code).

Conventions could include how these metrics and their thresholds should be provided, which attributes/resources they use or how a plugin execution is provided as an Event.

If this is a sensible proposal, I'd love to contribute.  

References:

1. https://nagios-plugins.org/doc/guidelines.html#AEN200 
2. https://www.monitoring-plugins.org/ 
3. https://github.com/Linuxfabrik/monitoring-plugins 
4. https://docs.checkmk.com/latest/en/localchecks.html#metrics 
5. https://icinga.com/docs/icinga-2/latest/doc/05-service-monitoring/#performance-data-metrics 

Slack Threat: https://cloud-native.slack.com/archives/C07UWG2ATQB/p1764329658484799 

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

**Comment by thompson-tomo** [comment]
So my initial take on this would be that:

- An open telemetry collector reciever/processor is written to recieve the nagios metrics and transform them to otel equivalents. This transformation would
  - take the label and resolve the otel metric name, description, type & fixed attributes.
  - if min/max are specified and non 0, additional complementary metrics are created to capture that just like we do with memory etc
  - if warn/crit are specified and non 0, a complementary status metric is used to capture the state
- we capture the nagios check as a `health_check` event as discussed in #1106.

**Comment by martialblog** [comment]
> An open telemetry collector reciever/processor is written to recieve the nagios metrics and transform them to otel equivalents. 

Yes I think that makes sense. Tools like Telegraf already do something similar. https://github.com/influxdata/telegraf/tree/release-1.36/plugins/parsers/nagios 

> we capture the nagios check as a health_check event

Interesting, I'll have a look at that proposal. I did search for similar open issues but the terminology used in the "Nagios-compatible monitoring realm" is sometimes tricky, since every tool or vendor calls the same thing differently (checks, status checks, health checks, and so on). 

There's not even a good term for "Nagios-compatible monitoring tool". I think, Nagios was just the first tool that used return codes and the output format and other tools are either forks or inspired by it. Unfortunately there is no actual standard for this either.

**Comment by thompson-tomo** [comment]
Agree hard part is coming up with naming when you don't have a standard to follow.

**Comment by martialblog** [comment]
Hi,

I've been thinking about this some more.

Since each of these results contains multiple metrics and thus it should be split up in an
expected manor. For example:

`DISK CRITICAL - free space: / 13969 MB (77% inode=99%); /boot 2 MB (0% inode=96%);
| /=4151MB;14496;16308;0;18121 /boot=294MB;236;266;0;296`

The actual result of the check:

- result - 0,1,2,3 (for OK, WARNING, CRITICAL, UNKNOWN and so on)

However, maybe this could also be expressed as a state metric: https://opentelemetry.io/docs/specs/semconv/how-to-write-conventions/status-metrics/

I'm not sure about this yet.

For each "performance data" bundle, a separate metric for each element is created
each of which MUST contain a "label" attribute for the respective label.

- performance_data
- performance_data_warn
- performance_data_crit
- performance_data_min
- performance_data_max

While the warn,crit,min,max values could also just be attributes on a single metric,
which would keep the "bundle" together, I think it makes more sense to have them be their own metric.

Mainly because so they can be queried (and represented in a chart) but also because of their cardinality it's probably not a good idea to have them be attributes.

For the above example that would be (using Prometheus notation just because I'm familiar with it):

- `performance_data{label="/", unit="MB"} 4151`
- `performance_data_warn{label="/"} 14496`
- `performance_data_crit{label="/"} 16308`
- `performance_data_min{label="/"} 0`
- `performance_data_max{label="/"} 18121`
- `performance_data{label="/boot", unit="MB"} 294`
- `performance_data_warn{label="/boot"} 236`
- `performance_data_crit{label="/boot"} 266`
- `performance_data_min{label="/boot"} 0`
- `performance_data_max{label="/boot"} 296`

Together with the standard attributes for resources (e.g. service.name, host.name, etc.) this should make it possible to store and query the specific results for any component.

Inspiration was taken from this prototype: https://codeberg.org/oxzi/icinga-prometheus-remote-writer

I think, the "health_check" or "state_check" proposal captures some of this,
but these "performance data" bundles are very specific to the "Nagios-compatible monitoring realm" and might need special treatment.

But I'm open to corrections and ideas.

**Comment by thompson-tomo** [comment]
So if I disect the below example

```
DISK CRITICAL - free space: / 13969 MB (77% inode=99%); /boot 2 MB (0% inode=96%); | /=4151MB;14496;16308;0;18121 /boot=294MB;236;266;0;296
```

I would extract the following
- `DISK CRITICAL - free space: / 13969 MB (77% inode=99%); /boot 2 MB (0% inode=96%);` check details
- `/=4151MB;14496;16308;0;18121` performance data # 1
- `/boot=294MB;236;266;0;296` performance data # 2

The check details are used in the health/state check metric as I assume there is a common pattern.

We use the data from a performance data as following:
- 4151: data point value for [`system.filesystem. utilization`](https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemfilesystemutilization) with [`system.filesystem.state`](https://opentelemetry.io/docs/specs/semconv/registry/attributes/system/) set to `free`
- `14496`: potentially not used
- `16308`: potentially not used
- `0`: not used in this case
- `18121`: data point value for [`system.filesystem.limit`](https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemfilesystemlimit)

All of these data points would have the label as an attribute.

I am still thinking as to the best way to capture the threshold but this could be attributes on the limit metric as done in https://opentelemetry.io/docs/specs/semconv/hardware/battery/#metric-hwbatterychargelimit but open to a nicer solution.

**Comment by martialblog** [comment]
Hi James, thanks for the feedback. You're dissection of the output is correct.

In an ideal world I agree that it would be best if each "performance data" would use a proper name for the metric name (e.g. `system.filesystem. utilization` as you described). However, I don't think that is feasible unfortunately. 

There's just a vast library of these "monitoring plugins" out there, written in every programming language conceivable. Some of these plugins have been in use for decades.

That means, that the tools/vendors using the plugins would have to insert the correct metric names, but they don't what plugins the users has applied. The "Nagios-compatible plugin interface" is the only agreement between the tools and their users. Also each tool/vendor would have to agree on which metric name to use for a given plugin.

This is why I currently think the best way to capture this ecosystem is to have a single (or set of) "performance data" metric that all vendors can agree on.

**Comment by thompson-tomo** [comment]
I agree on not wanting to change existing plugins however that is where the collector plugin would come in as it would transform the incoming data including changing the name and attributes to fit otel conventions. Hence avoids any need for changes in plugins. 

You would still have a generic health/state check plugin for high level as well as the event.

**Comment by dgoetz** [comment]
A challenge for the collector plugin trying to translate would be that the ecosystem is very wild and you can find monitoring plugins for very specific components including multiple modes with different performance data which might also vary. This is increased by the option to create your own plugins and I have seen this in most environments I have been involved as a consultant. 
Also the quality is varying so you will find some very loose interpretations of the standard. 

Just to give an example: The standard collection of plugins be it monitoring or nagios plugins does not contain a plugin to monitor memory usage on Linux which is a very basic metric you will want to monitor. So there are many attempts to fill this gap. One very old and common one is check_mem.pl which allows to query free or used memory, has an option to consider caches as free and all this changes the output including performance data. 
Because of this different vendors developed other plugins or included a mode in a plugin which does all the system monitoring. Just  to mention two NETWAYS has check_system_basics with mode memory which is Linux only and Linux Fabrik has memory_usage which covers Linux and Windows.

So having an ever growing mapping for translation which needs to be kept up to date and in sync in your environment while allowing your own expansion for self-written plugins, sounds at least for me not manageable. So a more simple generic approach sound more manageable, but of course I have to agree also like less quality.

**Comment by dgoetz** [comment]
And while thinking about I found another challenge as there is also an extension of the format that is not uncommon which is called multi format or check_multi format.

It originates from [check_multi](https://github.com/flackem/check_multi/blob/next/doc/configuration/performance.md) and is for example also used in [Icinga for Windows](https://icinga.com/docs/icinga-for-windows/latest/mssql/doc/plugins/01-Invoke-IcingaCheckMSSQLBackupStatus/).

This format is used to combine multiple checks in one execution and then one output, so it needs also to combine the performance data. 

When looking at the original format consisting of "service_description::plugin_name::performance_data" and how it is used by Icinga for Windows in the linked example with "database_name::plugin_name::performance_data" I am not even sure how wild this one can get.

(Sorry, for bringing up all of these challenges. But please do not shoot the messenger!)

**Comment by thompson-tomo** [comment]
So in the case for multi this should be split for otel.

In relation to the mapping I would do this via Configuration and not baked into the plugin. The other benefit of this would be it provides a way to offer enablement of the translation. Keep in mind I would still have supported a generic state/health event & potentially metric but it would not have the detailed performance data. If you want the performance data it needs to undergo a mapping.

**Comment by martialblog** [comment]
I agree that these "multi checks" should be split and then handled like any other result. There are probably more outliers like these we don't even know about. The closest thing there is to a spec/standard are the Nagios development docs, so that could be a common denominator.

---

I think the state result can fully be captured via the state/health_check proposal here: https://github.com/open-telemetry/semantic-conventions/issues/1106

Maybe with some optional additional attributes specific to the check plugins, not sure yet, depends on how flexible this enum is going to be `health_check.state -> enum` I guess? 

---

> In relation to the mapping I would do this via Configuration and not baked into the plugin.
> If you want the performance data it needs to undergo a mapping.

So my understanding was more that all the different vendors could at least agree upon a common format to capture this somewhat loose "Nagios-compatible monitoring plugin" stuff in one convention.

Since the vendors allow for users to add their own plugins, that configuration would mainly have to be provided by the user. 

Meaning, everyone would have to agree on a common "mapping" for each plugin out there, i.e. have a semantic convention that captures all plugins, so that vendors and users can have a guideline. 

Or am I wrong? Maybe my mental model of what OTel semantic convention are is wrong? 

That's why I suggested we provide a simple common ground for vendors to encode performance data metrics. "One convention to rule them all, and in the darkness bind them".

EDIT: maybe I should clarify, I'm not arguing against having a mapping when it's possible (e.g. vendors/tools MAY provide configuration for a mapping). I'm arguing for a generic common ground when it's not possible, which I think is going to happen often (tools/vendors MUST use the common format).

FYI, I've been doing to some quick and dirty prototyping here:

https://codeberg.org/martialblog/icinga-prometheus-remote-writer/src/branch/otlp

**Comment by d-riva** [comment]
> There's not even a good term for "Nagios-compatible monitoring tool". I think, Nagios was just the first tool that used return codes and the output format and other tools are either forks or inspired by it. Unfortunately there is no actual standard for this either.

@martialblog when I'm writing a new plugin, I follow this as my standard: https://www.monitoring-plugins.org/doc/guidelines.html

**Comment by RincewindsHat** [comment]
I am unsure about the relevance here, but I once attempted to write [a better description of the interface](https://datatracker.ietf.org/doc/html/draft-kaestle-monitoring-plugins-interface-04) which sadly did not make it, but might be interesting nonetheless.

**Comment by d-riva** [comment]
Special care could be warranted to handle the exotic format of the warning and critical threshold ranges.

Table 3. Example ranges
```
+-------------+---------------------------------------------+
| Range       | Generate an alert if x...                   |
+-------------+---------------------------------------------+
| 10          | < 0 or > 10 (outside {0 .. 10})             |
| 10:         | < 10 (outside {10 .. ∞})                    |
| ~:10        | > 10 (outside {-∞ .. 10})                   |
| 10:20       | < 10 or > 20 (outside {10 .. 20})           |
| @10:20      | ≥ 10 and ≤ 20 (inside {10 .. 20})           |
+-------------+---------------------------------------------+
```
Table 4. Command line examples


```
+--------------------------------------+-----------------------------------------------------------+
| Command line                         | Meaning                                                   |
+--------------------------------------+-----------------------------------------------------------+
| check_stuff -w10 -c20                | Critical if "stuff" > 20, else warn if > 10               |
|                                      | (will be critical if "stuff" < 0)                         |
| check_stuff -w~:10 -c~:20            | Same as above. Negative "stuff" is OK                     |
| check_stuff -w10: -c20               | Critical if "stuff" > 20, else warn if < 10               |
|                                      | (will be critical if "stuff" < 0)                         |
| check_stuff -c1:                     | Critical if "stuff" < 1                                   |
| check_stuff -w~:0 -c10               | Critical if "stuff" > 10; Warn if > 0                     |
|                                      | (will be critical if "stuff" < 0)                         |
| check_stuff -c5:6                    | Critical if "stuff" < 5 or > 6                            |
| check_stuff -c@10:20                 | OK if stuff < 10 or > 20, otherwise critical              |
+--------------------------------------+-----------------------------------------------------------+
```

**Comment by martialblog** [comment]
@d-riva Thanks for the hint. For this specific topic we are mainly interested in the output of the plugins and how it should be named and what meaning it has.

**Comment by d-riva** [comment]
@martialblog the ranges will show up in the perfdata warn and crit fields.

- `performance_data_warn{label="/boot"} 236` is a shorthand for the range `0:236`
- `performance_data_crit{label="/boot"} 266` is a shorthand for the range `0:266`

Also if a range is used in the command line to run a plugin it's highly probable that it will show up in the same form in the perfdata output.

**Comment by thompson-tomo** [comment]
> depends on how flexible this enum is going to be health_check.state -> enum I guess?

Enums are open so no restrictions.

> That's why I suggested we provide a simple common ground for vendors to encode performance data metrics.

The problem I see is that a generic performance data metric would be too broad and usefulness would be limited when compared to a transformed metric. Perhaps a good example of this would be prometheus which does this transformation approach.

Note I haven incorporated threshold & limit metrics into the generic health/state check issue

**Comment by martialblog** [comment]
@d-riva Wow, I did not know that! Been working with this stuff for years now and there's always something new. Thanks

So far, I've not seen other output than a float/int for the performance data metrics. Would be interesting to see how different tools/vendors handle this.

**Comment by martialblog** [comment]
@thompson-tomo 

> The problem I see is that a generic performance data metric would be too broad and usefulness would be limited when compared to a transformed metric. 

I do agree with that.

With a generic performance data metric all those "monitoring plugin" metrics will live if their own little world. When a "check_disk" uses "system.filesystem.usage" instead of a generic metric, its output can be connected to all other  "system.filesystem.usage" metrics.  Which is the intention if this semantic convention: 

>  The benefit to using Semantic Conventions is in following a common naming scheme that can be standardized across a codebase, libraries, and platforms. This allows easier correlation and consumption of data.

However, maybe one use of a generic metric could be some sort of "fallback" when mapping isn't possible/configured at the producing side (Icinga, Naemon, Nagios, etc.).

Then a later component (e.g. a collector) can easily identify them and try to enrich/map them if possible. 
Otherwise vendors might start to send stuff with their vendors-specific naming and a collector receiver/processor implementation is more complex because it first has to be aware of the vendor. You know what I mean?

Let's say Icinga, Naemon, Nagios (which do essentially the same thing) send performance data metrics like this:

```bash
# output of "check_foo"
OK - check_foo | metricA=1;

# each tool starts encoding things however they see fit

icinga_performance_data_metric{service.name=example, check=check_foo, label=metricA} 1

performance_data_metric_nagios{service.name=example, check.name=check_foo, data_label=metricA} 1

naemon_check_metric{service.name=example, naemon.chk=check_foo, metric.label=metricA} 1
```

This is basically what I want to avoid, since all three execute the same plugin and it should at least have the same format/naming. Yes, the actual semantic of whatever check_foo expresses is maybe not caught properly, but at least we can envision a component in the processing pipeline that can translate it. 

> Perhaps a good example of this would be prometheus which does this transformation approach.

Do you mean this? https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/translator/prometheus 

If not, could you point me to where I can read up on what transformation Prometheus does.

**Comment by thompson-tomo** [comment]
> However, maybe one use of a generic metric could be some sort of "fallback" when mapping isn't possible/configured at the producing side (Icinga, Naemon, Nagios, etc.).

For that I would use the state/health check event as mentioned earlier. The event has a body which can contain the original result hence no need for agents to include any attributes outside of the check name & result. 

I could see the following end to end use case.

1. Nagios agent runs check
2. Check result is logged ideally as a structured log.
3. Log is transferred to collector ideally directly via otlp but indirectly ie file also works.
4. Collector pipeline runs and based on nagios transformer config decides if specialised metric is used for performance data or just general metric/s.
5. Collector emits the original event plus any metrics.

With this approach in an ideal world agents export logs directly via otlp but in worse case 1 or more reciever are needed to support the agents not supporting otlp. But we avoid the overall generic metrics, keep things simple for the nagios agents & ensure all data from the check is available for processing

**Comment by d-riva** [comment]
> So far, I've not seen other output than a float/int for the performance data metrics. Would be interesting to see how different tools/vendors handle this.

They're actually strings as they can contain an optional unit of measurement. 'label'=value[UOM];[warn];[crit];[min];[max]

Here the documentation on how performance data metrics are handled in Icinga2: https://icinga.com/docs/icinga-2/latest/doc/05-service-monitoring/#performance-data-metrics and how the parsed metrics get exported again https://icinga.com/docs/icinga-2/latest/doc/14-features/#metrics
